File: internal/infrastructure/media/ffmpeg/adminmedia.go
package ffmpeg

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/Marcuss-ops/PipelineGen/internal/application/adminmedia"
	"github.com/Marcuss-ops/PipelineGen/internal/infrastructure/process"
)

// AdminMediaProcessor adapts the canonical FFmpeg Processor to admin-media
// application ports. Processor is optional for backwards-compatible zero
// construction; production should use NewAdminMediaProcessor.
type AdminMediaProcessor struct {
	Processor *Processor
}

func NewAdminMediaProcessor(ffmpegPath string) AdminMediaProcessor {
	if strings.TrimSpace(ffmpegPath) == "" {
		ffmpegPath = "ffmpeg"
	}
	return AdminMediaProcessor{Processor: NewProcessor(ffmpegPath)}
}

func (p AdminMediaProcessor) processor() *Processor {
	if p.Processor != nil {
		return p.Processor
	}
	return NewProcessor("ffmpeg")
}

func (p AdminMediaProcessor) Probe(ctx context.Context, path string) (time.Duration, error) {
	info, err := p.processor().Probe(ctx, path)
	if err != nil {
		return 0, err
	}
	return info.Duration, nil
}

func (p AdminMediaProcessor) Trim(ctx context.Context, inputPath string, maxSeconds float64) error {
	proc := p.processor()
	ext := strings.ToLower(filepath.Ext(inputPath))
	tmpPath := inputPath + ".trim.tmp" + ext
	defer os.Remove(tmpPath)
	args := []string{"-y", "-i", inputPath, "-t", fmt.Sprintf("%.3f", maxSeconds)}
	switch ext {
	case ".mp4", ".mov", ".mkv":
		args = append(args, "-map", "0:v:0?", "-map", "0:a:0?", "-c:v", "libx264", "-preset", "medium", "-crf", "18", "-c:a", "aac", "-b:a", "192k", "-movflags", "+faststart")
	case ".wav":
		args = append(args, "-vn", "-c:a", "pcm_s16le")
	default:
		args = append(args, "-vn", "-c:a", "libmp3lame", "-q:a", "2")
	}
	args = append(args, tmpPath)
	if _, err := proc.runner.Run(ctx, proc.path, args, process.Options{Timeout: 10 * time.Minute, CombinedOutput: true}); err != nil {
		return fmt.Errorf("ffmpeg trim: %w", err)
	}
	return os.Rename(tmpPath, inputPath)
}

func (p AdminMediaProcessor) Render(ctx context.Context, m adminmedia.RenderManifest) error {
	ff := []string{"-y", "-i", m.Input}
	for _, e := range m.Effects {
		ff = append(ff, "-i", e.Path)
	}
	video := "[0:v]"
	for _, o := range m.Overlays {
		color, y := o.Color, o.Y
		if color == "" {
			color = "white"
		}
		if y == "" {
			y = "h*0.50"
		}
		text := strings.ReplaceAll(o.Text, "'", "\\\\'")
		video += fmt.Sprintf("drawtext=fontfile=%s:text='%s':fontcolor=%s:fontsize=%s:borderw=3:bordercolor=black:x=(w-text_w)/2:y=%s:enable='between(t\\,%s\\,%s)',", m.Font, text, color, o.Size, y, o.Start, o.End)
	}
	video = strings.TrimSuffix(video, ",") + "[vout]"
	filter := video + ";[0:a]aresample=48000,volume=0.78[base]"
	labels := "[base]"
	for i, e := range m.Effects {
		duration := e.Duration
		if duration <= 0 {
			duration = 1
		}
		volume := e.Volume
		if volume == "" {
			volume = "1"
		}
		filter += fmt.Sprintf(";[%d:a]aresample=48000,atrim=duration=%.3f,asetpts=PTS-STARTPTS,volume=%s,adelay=%d|%d[sfx%d]", i+1, duration, volume, e.DelayMS, e.DelayMS, i)
		labels += fmt.Sprintf("[sfx%d]", i)
	}
	filter += fmt.Sprintf(";%samix=inputs=%d:duration=first:dropout_transition=0:normalize=0,alimiter=limit=0.95[aout]", labels, len(m.Effects)+1)
	ff = append(ff, "-filter_complex", filter, "-map", "[vout]", "-map", "[aout]", "-c:v", "libx264", "-preset", "medium", "-crf", "18", "-c:a", "aac", "-b:a", "192k", "-movflags", "+faststart", "-shortest", m.Output)
	proc := p.processor()
	_, err := proc.runner.Run(ctx, proc.path, ff, process.Options{Timeout: 20 * time.Minute, CombinedOutput: true})
	return err
}

var _ adminmedia.AudioEditor = AdminMediaProcessor{}
var _ adminmedia.ShortRenderer = AdminMediaProcessor{}

File: rust/pipelinegen-muscles/src/admin_media.rs
use crate::artifact::{failed_response, part_path, publish_output};
use crate::encoder::append_video_options;
use crate::protocol::{Request, Response};
use std::fs;
use std::path::Path;
use std::process::Command;

pub(crate) fn execute(request: Request) -> Response {
    admin_render(request)
}

fn admin_render(request: Request) -> Response {
    let input = match request.source_path.as_deref() {
        Some(path) if Path::new(path).is_file() => path,
        _ => {
            return failed_response(
                request.source_path.clone(),
                "source_path is required and must be readable".to_string(),
            )
        }
    };
    let output = match request.output_path.as_deref() {
        Some(path) if !path.is_empty() => path,
        _ => {
            return failed_response(
                Some(input.to_string()),
                "output_path is required".to_string(),
            )
        }
    };
    let font = match request.font.as_deref() {
        Some(path) if !path.is_empty() => path,
        _ => return failed_response(Some(input.to_string()), "font is required".to_string()),
    };
    let profile = match request.media.profile() {
        Ok(value) => value,
        Err(error) => return failed_response(Some(input.to_string()), error),
    };
    let effects = request.effects.as_ref().cloned().unwrap_or_default();
    let overlays = request.overlays.as_ref().cloned().unwrap_or_default();
    let part = part_path(output);
    let mut command = Command::new(request.ffmpeg_path.as_deref().unwrap_or("ffmpeg"));
    command.args(["-hide_banner", "-loglevel", "error", "-y", "-i", input]);
    for effect in &effects {
        if !Path::new(&effect.path).is_file() {
            return failed_response(
                Some(input.to_string()),
                format!("effect is not readable: {}", effect.path),
            );
        }
        command.args(["-i", &effect.path]);
    }
    if overlays.is_empty() {
        return failed_response(
            Some(input.to_string()),
            "at least one overlay is required".to_string(),
        );
    }
    let mut video = String::from("[0:v]");
    for overlay in overlays {
        let color = if overlay.color.is_empty() {
            "white"
        } else {
            &overlay.color
        };
        let y = if overlay.y.is_empty() {
            "h*0.50"
        } else {
            &overlay.y
        };
        let text = overlay.text.replace('\'', "\\\\'");
        video.push_str(&format!("drawtext=fontfile={}:text='{}':fontcolor={}:fontsize={}:borderw=3:bordercolor=black:x=(w-text_w)/2:y={}:enable='between(t\\,{}\\,{})',", font, text, color, overlay.size, y, overlay.start, overlay.end));
    }
    video = format!("{}[vout]", video.trim_end_matches(','));
    let mut filter = format!("{};[0:a]aresample={},volume=0.78[base]", video, profile.sample_rate);
    let mut labels = String::from("[base]");
    for (index, effect) in effects.iter().enumerate() {
        let duration = if effect.duration <= 0.0 {
            1.0
        } else {
            effect.duration
        };
        let volume = if effect.volume.is_empty() {
            "1"
        } else {
            &effect.volume
        };
        filter.push_str(&format!(";[{}:a]aresample={},atrim=duration={:.3},asetpts=PTS-STARTPTS,volume={},adelay={}|{}[sfx{}]", index + 1, profile.sample_rate, duration, volume, effect.delay_ms, effect.delay_ms, index));
        labels.push_str(&format!("[sfx{}]", index));
    }
    filter.push_str(&format!(";{}amix=inputs={}:duration=first:dropout_transition=0:normalize=0,alimiter=limit=0.95[aout]", labels, effects.len() + 1));
    command.args([
        "-filter_complex",
        &filter,
        "-map",
        "[vout]",
        "-map",
        "[aout]",
    ]);
    if let Err(error) = append_video_options(&mut command, &request) {
        return failed_response(None, error);
    }
    command.args([
        "-c:a",
        profile.audio_codec.as_str(),
        "-b:a",
        profile.audio_bitrate.as_str(),
        "-ar",
        &profile.sample_rate.to_string(),
        "-ac",
        &profile.channels.to_string(),
        "-movflags",
        "+faststart",
        "-shortest",
        &part,
    ]);
    match command.output() {
        Ok(result) if result.status.success() => match publish_output(&part, output) {
            Ok(()) => Response {
                ok: true,
                operation: "admin_render".to_string(),
                source_path: Some(input.to_string()),
                items: Vec::new(),
                metadata: None,
                error: None,
            },
            Err(error) => failed_response(Some(input.to_string()), error),
        },
        Ok(result) => {
            let _ = fs::remove_file(&part);
            failed_response(
                Some(input.to_string()),
                format!(
                    "admin render failed: {}",
                    String::from_utf8_lossy(&result.stderr).trim()
                ),
            )
        }
        Err(error) => {
            let _ = fs::remove_file(&part);
            failed_response(
                Some(input.to_string()),
                format!("admin render failed: {error}"),
            )
        }
    }
}
